src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -94,62 +103,76 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
+        // No worker available, we won't be able to schedule any kind of task.
+        return;
+      }
 
-    std::shared_ptr<TaskResourceInstances> allocated_instances(
-        new TaskResourceInstances());
-    bool schedulable = cluster_resource_scheduler_->AllocateLocalTaskResources(
-        spec.GetRequiredResources().GetResourceMap(), allocated_instances);
-    if (!schedulable) {
-      // Not enough resources to schedule this task.
-      // Put it back at the end of the dispatch queue.
-      tasks_to_dispatch_.push_back(work);
-      worker_pool.PushWorker(worker);
-      // Try next task in the dispatch queue.
-      continue;
-    }
+      std::shared_ptr<TaskResourceInstances> allocated_instances(
+          new TaskResourceInstances());
+      bool schedulable = cluster_resource_scheduler_->AllocateLocalTaskResources(
+          spec.GetRequiredResources().GetResourceMap(), allocated_instances);
+      if (!schedulable) {
+        // Not enough resources to schedule this task.
+        worker_pool.PushWorker(worker);
+        // All the tasks in this queue are the same, so move on to the next queue.
+        break;
+      }
 
-    auto reply = std::get<1>(work);
-    auto callback = std::get<2>(work);
-    worker->SetOwnerAddress(spec.CallerAddress());
-    if (spec.IsActorCreationTask()) {
-      // The actor belongs to this worker now.
-      worker->SetLifetimeAllocatedInstances(allocated_instances);
-    } else {
-      worker->SetAllocatedInstances(allocated_instances);
+      auto reply = std::get<1>(work);
+      auto callback = std::get<2>(work);
+      worker->SetOwnerAddress(spec.CallerAddress());
+      if (spec.IsActorCreationTask()) {
+        // The actor belongs to this worker now.
+        worker->SetLifetimeAllocatedInstances(allocated_instances);
+      } else {
+        worker->SetAllocatedInstances(allocated_instances);
+      }
+      worker->AssignTaskId(spec.TaskId());
+      if (!RayConfig::instance().enable_multi_tenancy()) {
+        worker->AssignJobId(spec.JobId());
+      }
+      worker->SetAssignedTask(task);
+      Dispatch(worker, leased_workers, spec, reply, callback);
+      work_it = dispatch_queue.erase(work_it);
     }
-    worker->AssignTaskId(spec.TaskId());
-    if (!RayConfig::instance().enable_multi_tenancy()) {
-      worker->AssignJobId(spec.JobId());
+    if (dispatch_queue.empty()) {
+      shapes_it = tasks_to_dispatch_.erase(shapes_it);
+    } else {
+      shapes_it++;
     }
-    worker->SetAssignedTask(task);
-    Dispatch(worker, leased_workers, spec, reply, callback);
   }
 }
 
 void ClusterTaskManager::QueueTask(const Task &task, rpc::RequestWorkerLeaseReply *reply,
                                    std::function<void(void)> callback) {
   Work work = std::make_tuple(task, reply, callback);
-  tasks_to_schedule_.push_back(work);
+  const auto &scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
+  auto it = tasks_to_schedule_.find(scheduling_class);
```

Comment:
Minor, but I think you can do tasks_to_schedule_[scheduling_class].push_back in both cases.